src/web/razor/Program.cs

```diff
@@ -1,26 +1,121 @@
+
+builder.Services.AddPiranha(options =>
 {
-    public class Program
+    /*
+     * This will enable automatic reload of .cshtml
+     * without restarting the application. However since
+     * this adds a slight overhead it should not be
+     * enabled in production.
+     */
+    options.AddRazorRuntimeCompilation = true;
+
+    options.UseCms();
+    options.UseManager();
+
+#if (UseBlobStorage)
+    options.UseBlobStorage(builder.Configuration.GetConnectionString("blobstorage"), naming: Piranha.Azure.BlobStorageNaming.UniqueFolderNames);
+#else
+    options.UseFileStorage(naming: Piranha.Local.FileStorageNaming.UniqueFolderNames);
+#endif
+    options.UseImageSharp();
+    options.UseTinyMCE();
+    options.UseMemoryCache();
+
+    var connectionString = builder.Configuration.GetConnectionString("piranha");
+#if (UseSQLServer)
+    options.UseEF<SQLServerDb>(db => db.UseSqlServer(connectionString));
+    options.UseIdentityWithSeed<IdentitySQLServerDb>(db => db.UseSqlServer(connectionString));
+#elif (UseMySql)
+    options.UseEF<MySqlDb>(db => db.UseMySql(connectionString, ServerVersion.AutoDetect(connectionString)));
+    options.UseIdentityWithSeed<IdentityMySQLDb>(db => db.UseMySql(connectionString, ServerVersion.AutoDetect(connectionString)));
+#elif (UsePostgreSql)
+    options.UseEF<PostgreSqlDb>(db => db.UseNpgsql(connectionString));
+    options.UseIdentityWithSeed<IdentityPostgreSQLDb>(db => db.UseNpgsql(connectionString));
+#else
+    options.UseEF<SQLiteDb>(db => db.UseSqlite(connectionString));
+    options.UseIdentityWithSeed<IdentitySQLiteDb>(db => db.UseSqlite(connectionString));
+#endif
+
+    /*
+     * Here you can configure the different permissions
+     * that you want to use for securing content in the
+     * application.
+    options.UseSecurity(o =>
     {
-        public static void Main(string[] args)
-        {
-            CreateHostBuilder(args).Build().Run();
-        }
-
-        public static IHostBuilder CreateHostBuilder(string[] args) =>
-            Host.CreateDefaultBuilder(args)
-                .ConfigureWebHostDefaults(webBuilder =>
-                {
-                    webBuilder.UseStartup<Startup>();
-                });
-    }
+        o.UsePermission("WebUser", "Web User");
+    });
+     */
+
+    /*
+     * Here you can specify the login url for the front end
+     * application. This does not affect the login url of
+     * the manager interface.
+    options.LoginUrl = "login";
+     */
+});
+
+var app = builder.Build();
+
+// For application behind a reverse proxy, accept forwarded headers
+// Nginx: https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/linux-nginx?view=aspnetcore-6.0
+// Apache: https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/linux-apache?view=aspnetcore-6.0
+//app.UseForwardedHeaders(new ForwardedHeadersOptions
+//{
+//    ForwardedHeaders = ForwardedHeaders.All
+//});
+
+var api = app.Services.GetService<IApi>();

```

Comment:
See comment in MVC template

Reply:
I have tried both templates, see screenshots above. And I think the problem is not in calling a single GetService<>, but in a try to inject IApi which is scoped into a Singleton which doesn't allow scoped injections. I refer to https://stackoverflow.com/questions/48590579/cannot-resolve-scoped-service-from-root-provider-net-core-2
That's why I am asking for an explicit stack trace. The MoveNext() in stack trace shows that the issue is not in initialization (init is sync) but in await app.RunAsync(); as it is the only async instruction in the Program.cs.

But if we refer to Program.cs line 88, it is

```
84 // Configure Tiny MCE
85 EditorConfig.FromFile("editorconfig.json");
86
87 // Middleware setup
88 app.UsePiranha(options =>
89 {
90    options.UseManager();
91    options.UseTinyMCE();
92    options.UseIdentity();
93 });
```

That's why I asked @goranhell to attach his full version of Program.cs and stack trace simultaneously.

Reply:
Please see my other comment.